### PR TITLE
Breaking: Error thrown/printed if no config found (fixes #5987)

### LINKED ISSUE #5987: Proposal: throw error if no config file is found
From my travels around the internet, it seems that first-time ESLint users are very confused about the "nothing by default" mode. It seems many do not read the getting started docs, and so don't know to use `--init` to get started. That leaves them thinking that ESLint is broken because JSHint shows output out of the box. 

~~To remedy this, I'd like to propose that in 3.0, we change the default behavior to use `eslint:recommended ` when we don't find a config file. This should only affect CLI users and not affect `CLIEngine` by default (so as not to affect API consumers).~~

~~Another option is to automatically run `--init` when a config file isn't found, but that won't work well in editors, where an interactive prompt won't be available.~~

**Edit:** new proposal is throwing an error asking the user to run --init (we can use a pretty error message via messagsTemplate rather than showing the stack trace).

### PR DESCRIPTION
**Pull request by IanVS**
Continues from #6384.

This adds a few tests to @platinumazure's PR.  That PR was from his own branch, and I couldn't figure out a way to add to his PR.

Unfortunately, one of the tests I added, at the suggestion from @mysticatea in https://github.com/eslint/eslint/pull/6384#issuecomment-225448030, failed.  So I needed to add one more commit to prevent that.  Please let me know if there are any problems with this approach, @platinumazure.

### CODE REVIEW

File: lib/config.js [hunk 1]
```diff
@@ -154,6 +154,8 @@ function getLocalConfig(thisConfig, directory) {
 
         if (personalConfig) {
             return ConfigOps.merge(config, personalConfig);
+        } else if (thisConfig.options.rules && thisConfig.options.rules !== {}) {
```
Comment (nzakas) [inline lib/config.js]: This comparison will always return `true`. The `{}` creates an entirely new object, so `thisConfig.options.rules` can never equal `{}`. What is it you are trying to test for here?

Comment (IanVS) [inline lib/config.js]: Doh, I wondered about that.  I want to check that it's not empty.  So maybe `Object.keys(thisConfig.options.rules).length > 0` would work?

Comment (IanVS) [inline lib/config.js]: Sorry, I should have been more careful.  I'm rushing and that's never a good thing.

Comment (nzakas) [inline lib/config.js]: Yeah, that's a better way to do it. Also, can you add in a comment explaining when this condition will be hit?

**Comment by platinumazure** [comment]
@IanVS Sorry, if I knew you were going to add stuff I would have made you a contributor on my fork. I'll take a look and see what you've added here...

**Comment by platinumazure** [comment]
@IanVS Actually I'm not sure about the `--rule` use case. I might have been too quick to accept that as a suggested test. Let me explain:

* `eslint . --rule semi:error` should combine the override rule setting with an ESLint config, so I would actually still expect this to throw if no config is found.
* `eslint . --no-eslintrc --rule semi:error` on the other hand should not throw if no config is found.

Based on that reasoning, I think your second unit test actually should expect a throw, and that a third test should be created which has `useEslintrc: false` and a rule override and that should not throw.

What do you think?

**Comment by nzakas** [comment]
@platinumazure I disagree with your first case. Passing in `--rule` is the same as having a configuration file. I don't think that should throw.

**Comment by nzakas** [comment]
Actually, we might want to expand that to `--env`, `--global`, and `--plugin`, as well. Basically, if you're providing configuration options on the command line, I think that should count as having a configuration.

Remember: our goal here is not to enforce a configuration *file* for each run, just to enforce a *configuration*.

**Comment by platinumazure** [comment]
@nzakas Ack, you might be right. I guess that means lib/config.js needs to be updated to see if any of those options were passed in, and not throw in those cases?

**Comment by nzakas** [comment]
@platinumazure yes.

**Comment by platinumazure** [comment]
@nzakas Do we want to check for presence of the options, or presence of actual keys/values in the option values? (In other words, if someone sends in `rules: {}`, should that be considered good enough?

**Comment by nzakas** [comment]
@platinumazure sorry, I don't understand the question.

**Comment by platinumazure** [comment]
@nzakas I'm asking about whether we want to differentiate between these two cases:

```js
new CLIEngine({
    rules: {}
});

// vs

new CLIEngine({
    // rules: undefined
});
```

Neither of those represents having "a configuration", arguably. (I omitted the other options for brevity, but basically I'm asking if we should only count a consumer as "having provided a configuration" if they actually have a value in at least one of rules/env/plugins/globals (and for that matter, what about parser and parserOptions?).

**Comment by nzakas** [comment]
@platinumazure ah gotcha, thanks for explaining. I think we want to be sure that *something* is configured. So `rules: {}` should still throw if there's nothing else in `env` or `globals`.

Thinking about it some more, plugins probably shouldn't count because they can't do anything by default. 

For parser and parserOptions...hmmm, that's trickier. My initial inclination is that they shouldn't count when provided on their own (because you're not doing any linting at that point -- still just parsing).

**Comment by IanVS** [comment]
I'd say there's no difference between your two cases, which is why I explicitly checked that the length of the rules object is greater than zero.  The first case corresponds to something like `eslint --rule` with no arguments, which isn't a configuration at all.

I agree that we should check for `--plugin` as well, but I think that's all.  `--parser` doesn't really do any good without specifying something else like a rule as well.  Same with `--env` and `--global`, I believe there won't be any errors regardless of those settings unless `no-undef` is enabled (although I could be wrong there. . . ).

**Comment by nzakas** [comment]
@ianvs I revised my opinion and don't think we should count `--plugin` as having a configuration because it doesn't do anything unless you end up using the plugin's rules, environment, or configuration.

**Comment by platinumazure** [comment]
@nzakas If our goal is to throw if we don't have "a configuration" rather than "a configuration file", should we also throw if we did in fact find a configuration file, but said file has no rules?

And is it sufficient to check `rules`, since `env` and `globals` don't really have any effect except through rules? (The parser doesn't try to detect if a global-scope reference matches our globals, right?)

**Comment by nzakas** [comment]
@platinumazure hmmm... initially I was thinking a config file with no rules would be ok, because someone might want to just check parsing, and the only other way to do that would be to use `--no-eskintrc`, which would be ok for the default parser in ES5, but a pain for non-default settings. So I still think a config file, even if it does have rules, should not throw an error. 

Good point about env and globals. I think you're correct in that rules is the only thing we care about.

**Comment by platinumazure** [comment]
Okay, so I'll detect that we have a config file (or custom config file
location that exists, or --no-eslintrc), or else that we have at least one
rule; and if we have none of those, we throw.

How does that sound?

**Comment by nzakas** [comment]
Sounds great!

**Comment by IanVS** [comment]
@platinumazure, I believe what you described is the current state of this PR, right?

**Comment by platinumazure** [comment]
@IanVS I think so but I'm not certain. At any rate I want to reorganize the
code a little bit. I've fast-forwarded my branch to yours and can add
another couple of commits tomorrow- is it okay if I just do that and create
a new PR tomorrow? Thanks for all your help!

**Comment by IanVS** [comment]
Yep, fine with me! Take it away.

**Comment by IanVS** [comment]
Superseded by https://github.com/eslint/eslint/pull/6538.

**Pull request closed without merging**